### ISSUE TITLE
feat(commands/apm/publish): introduce option to perform version bump

### ISSUE THREAD

**Issue by PascalPrecht** [aragon/aragon-cli#195]
The development setup as it is right now, may introduce an out of sync state between the version of a contract deployed on the chain and the version stored in an app's `arapp.json` file. 

**Here's one scenario**:

To develop an aragon sandboxed app locally, we use

```
$ aragon run
```

which will spin up a devchain, deploy the app's contracts and also increases the version of the deployed contract by one major version. In other words, running `aragon run` 5 times in a row, will result in the app's contract being deployed 5 times, with 5 versions, `1.0.0`, `2.0.0`, `3.0.0`, `4.0.0`, `5.0.0` respectively.

Notice that this happens locally on a devchain which can be reset any time.

However, when running

```
$ aragon apm publish
```

which **also** deploys the app to the configured chain, CLI is actually reading the version number stored in a sandboxed app's `arapp.json` file and checks whether that version is a valid "next" version on the chain or not. 

So assuming 5.0.0 of our app contract is deployed on chain, according to aragon's versioning rules, valid next versions would be: `5.0.1`, `5.1.0` or `6.0.0`.

Since the version number in `arapp.json` is not updated when running `aragon run` it's still on `1.0.0`, resulting in the CLI throwing an error that 1.0.0 is not a valid version to bump to.

Once can use

```
$ aragon apm version [patch|minor|major]
```

to increase that version and try again. 

The problem still exists though, if we don't run `aragon apm version major` 5 times in a row. Obviously, another option would be to just change the version in `arapp.json` manually.

While those are solutions to the problem, this is a very confusing setting especially if developers aren't familiar with the inner workings of aragon's publishing system. Ideally, version numbers wouldn't get out of sync in the first place and would be retrieved from a single source of truth.

**Solution to this problem**

Instead of relying on the version number stored in `arapp.json`, we can tell `apm publish` to simply take the latest version number from the chain and then increase it by either patch, minor or major. This means we'd need a new option to specify the version bump type (naming to be discussed):

```
$ aragon apm publish --version-bump=minor
```

This would also make the version number stored in `arapp.json` obsolete. However, at the same time, this would be a breaking change if we required users to specify the `--version-bump`. So before `version` property in `arapp.json` gets dropped and `--version-bump` is a requirement, we can support both scenarios.

If no `--version-bump` is given, we would still go ahead and read the version from `arapp.json`. In case the numbers are out of sync, we can improve the error message that `--version-bump` can be used.

**Comment by izqui** [comment]
I like the proposed solution.

However if we introduce this way of performing version bumps that takes the on-chain last published version as the base version to perform the bump, we should also change `aragon apm version [bump]` so it no longer does the bump based on the version in the `arapp.json` file.

It would be very confusing to have `aragon apm publish --version-bump=[bump]` and `aragon apm version [bump]` behave in completely different ways.

**Comment by PascalPrecht** [comment]
Agreed. Created https://github.com/aragon/aragon-cli/issues/196 accordingly

**Comment by sohkai** [comment]
> This would also make the version number stored in arapp.json obsolete.

I think it'd still be useful for a generated file to contain the `version` field, that's used during publishing, so it's easy to verify a particular deployment.

The tooling could add this into the file into its tmp dir when deploying.

**Comment by PascalPrecht** [comment]
> I think it'd still be useful for a generated file to contain the version field, that's used during publishing, so it's easy to verify a particular deployment.

Yes. For now we went with a hybrid solution as well. The version is always synced back into `arapp.json`, using the on chain version as single source of truth. 

Whether we want to get rid of the version field eventually or not can then be revisited.
